mesop/web/src/services/channel.ts

```diff
@@ -117,143 +139,204 @@ export class Channel {
-              .getStateToken()!;
-            switch (uiResponse.getUpdateStateEvent()!.getTypeCase()) {
-              case UpdateStateEvent.TypeCase.FULL_STATES: {
-                this.states = uiResponse
-                  .getUpdateStateEvent()!
-                  .getFullStates()!;
-                break;
-              }
-              case UpdateStateEvent.TypeCase.DIFF_STATES: {
-                const states = uiResponse
-                  .getUpdateStateEvent()!
-                  .getDiffStates()!;
-
-                const numDiffStates = states.getStatesList().length;
-                const numStates = this.states.getStatesList().length;
-
-                if (numDiffStates !== numStates) {
-                  throw Error(
-                    `Number of diffs (${numDiffStates}) doesn't equal the number of states (${numStates}))`,
-                  );
-                }
-
-                // `this.states` should be populated at this point since the first update
-                // from the server should be the full state.
-                for (let i = 0; i < numDiffStates; ++i) {
-                  const state = applyStateDiff(
-                    this.states.getStatesList()[i].getData() as string,
-                    states.getStatesList()[i].getData() as string,
-                  );
-                  this.states.getStatesList()[i].setData(state);
-                }
-                break;
-              }
-              case UpdateStateEvent.TypeCase.TYPE_NOT_SET:
-                throw new Error('No state event data set');
-            }
+        console.debug('Server event (SSE): ', uiResponse.toObject());
+        this.handleUiResponse(request, uiResponse, initParams);
+      });
+    });
+  }
+
+  private initWebSocket(initParams: InitParams, request: UiRequest) {
+    if (this.socket) {
+      this.status = ChannelStatus.OPEN;
+      const payload = generatePayloadString(request);
+      this.socket.emit('message', payload);
+      return;
+    }
+    this.socket = io('/__ui__', {
+      transports: ['websocket'],
+      reconnectionAttempts: 3,
+    });
+
+    this.status = ChannelStatus.OPEN;
+    this.isWaitingTimeout = setTimeout(() => {
+      this.isWaiting = true;
+    }, WAIT_TIMEOUT_MS);
+
+    this.logger.log({type: 'StreamStart'});
+
+    const {zone, onRender, onError, onCommand} = initParams;
+    this.initParams = initParams;
+
+    this.socket.on('connect', () => {
+      // Send the initial UiRequest upon connection
+      const payload = generatePayloadString(request);
+      this.socket!.emit('message', payload);
+    });
+
+    this.socket.on('response', (data: any) => {
+      const prefix = 'data: ';
+      const payloadData = (data.data.slice(prefix.length) as string).trimEnd();
+      zone.run(() => {
+        if (payloadData === STREAM_END) {
+          this._isHotReloading = false;
+          this.status = ChannelStatus.CLOSED;
+          this.logger.log({type: 'StreamEnd'});
```

Comment:
[question] What's the difference between using logger and console?

Reply:
logger is displayed in the Mesop DevTools - although TBH, I never really use it and I just use Chrome DevTools - I'm open to just ripping out logger & maybe the entire DevTools?